llvm/lib/CheerpWriter/CheerpWriter.cpp

```diff
@@ -1653,8 +1653,10 @@ void CheerpWriter::compileEqualPointersComparison(const llvm::Value* lhs, const
 	// NOTE: For any pointer-to-immutable, converting to CO is actually a dereference. (base[offset] in both cases)
 	//       PA enforces that comparisons between pointers-to-immutable (which include pointers-to-pointers)
 	//       need a SPLIT_REGULAR kind. Make sure to also use SPLIT_REGULAR if one kind is CONSTANT (e.g. null)
-	else if((lhsKind == REGULAR || lhsKind == SPLIT_REGULAR || lhsKind == RAW ||lhsKind == CONSTANT || (isGEP(lhs) && cast<User>(lhs)->getNumOperands()==2)) &&
-		(rhsKind == REGULAR || rhsKind == SPLIT_REGULAR || rhsKind == RAW ||rhsKind == CONSTANT || (isGEP(rhs) && cast<User>(rhs)->getNumOperands()==2)))
+	else if ((lhsKind == REGULAR || lhsKind == SPLIT_REGULAR || lhsKind == RAW ||lhsKind == CONSTANT ||
+		(isGEP(lhs) && (cast<User>(lhs)->getNumOperands()==2 || getGEPContainerType(cast<User>(lhs))->isArrayTy() || getGEPContainerType(cast<User>(lhs))->isPointerTy()))) &&
+			 (rhsKind == REGULAR || rhsKind == SPLIT_REGULAR || rhsKind == RAW ||rhsKind == CONSTANT ||
+		(isGEP(rhs) && (cast<User>(rhs)->getNumOperands()==2 || getGEPContainerType(cast<User>(rhs))->isArrayTy() || getGEPContainerType(cast<User>(rhs))->isPointerTy()))))
```

Comment:
I am not sure how the container type could ever be a pointer here. Also, please add a small lambda to check for this condition. The lambda can be used for both operands.

Reply:
right, we were thinking about including pointer type assuming that the utility function accepted geps with just one index too